waspc/src/Wasp/Analyzer/TypeChecker/Internal.hs

```diff
@@ -37,81 +37,83 @@ where
-checkStmt (P.Decl typeName name expr) =
+checkStmt :: P.WithCtx P.Stmt -> TypeChecker (WithCtx TypedStmt)
+checkStmt (P.WithCtx ctx (P.Decl typeName name expr)) =
   lookupDeclType typeName >>= \case
-    Nothing -> throw $ NoDeclarationType typeName
+    Nothing -> throw $ NoDeclarationType ctx typeName
     Just (TD.DeclType _ expectedType _) -> do
       -- Decides whether the argument to the declaration has the correct type
       mTypedExpr <- weaken expectedType <$> inferExprType expr
       case mTypedExpr of
-        Left e -> throw e
-        Right typedExpr -> return $ Decl name typedExpr (DeclType typeName)
+        Left e -> throw $ WeakenError ctx e
+        Right typedExpr -> return $ WithCtx ctx $ Decl name typedExpr (DeclType typeName)
 
 -- | Determine the type of an expression, following the inference rules described in
--- the wasplang document. Some these rules are referenced by name in the comments
+-- the wasplang document. Some of these rules are referenced by name in the comments
 -- of the following functions using [Brackets].
-inferExprType :: P.Expr -> TypeChecker TypedExpr
-inferExprType (P.StringLiteral s) = return $ StringLiteral s
-inferExprType (P.IntegerLiteral i) = return $ IntegerLiteral i
-inferExprType (P.DoubleLiteral d) = return $ DoubleLiteral d
-inferExprType (P.BoolLiteral b) = return $ BoolLiteral b
-inferExprType (P.ExtImport n s) = return $ ExtImport n s
-inferExprType (P.Var ident) =
-  lookupType ident >>= \case
-    Nothing -> throw $ UndefinedIdentifier ident
-    Just typ -> return $ Var ident typ
--- For now, the two quoter types are hardcoded here, it is an error to use a different one
--- TODO: this will change when quoters are added to "Analyzer.TypeDefinitions"
-inferExprType (P.Quoter "json" s) = return $ JSON s
-inferExprType (P.Quoter "psl" s) = return $ PSL s
-inferExprType (P.Quoter tag _) = throw $ QuoterUnknownTag tag
--- The type of a list is the unified type of its values.
--- This poses a problem for empty lists, there is not enough information to choose a type.
--- TODO: fix this in the future, probably by adding an additional phase to resolve type variables
---       that would be assigned here
-inferExprType (P.List values) = do
-  typedValues <- mapM inferExprType values
-  case unify <$> nonEmpty typedValues of
-    -- Apply [EmptyList]
-    Nothing -> return $ List [] EmptyListType
-    Just (Left e) ->
-      throw e
-    Just (Right (unifiedValues, unifiedType)) ->
-      return $ List (toList unifiedValues) (ListType unifiedType)
--- Apply [Dict], and also check that there are no duplicate keys in the dictionary
-inferExprType (P.Dict entries) = do
-  typedEntries <- zip (map fst entries) <$> mapM (inferExprType . snd) entries
-  dictType <-
-    foldM insertIfUniqueElseThrow M.empty $
-      second (DictRequired . exprType) <$> typedEntries
-  return $ Dict typedEntries (DictType dictType)
+inferExprType :: P.WithCtx P.Expr -> TypeChecker (WithCtx TypedExpr)
```

Comment:
In this file, in inferExprType, unify, unifyTypes, and weaken, I mostly deal with ensuring that ctx from the ASTs is correctly embedded into the generated errors. For that I needed to modify some of these functions so that they take more information as arguments than before.